Remotion/Web/Core/Context/HttpContextStorageProviderNetCore.cs

```diff
@@ -0,0 +1,188 @@
+// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Collections;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Web;
+using Remotion.Context;
+using Remotion.ServiceLocation;
+using Remotion.Utilities;
+
+namespace Remotion.Web.Context
+{
+#if !NETFRAMEWORK
+  [ImplementationFor(typeof(ISafeContextStorageProvider), Position = 0)]
+  public class HttpContextStorageProvider : ISafeContextStorageProvider, ISafeContextBoundaryStrategy
+  {
+    private class WrappedHashtable
+    {
+      public WrappedHashtable (HttpContext context, Hashtable data)
+      {
+        Context = context;
+        Data = data;
+      }
+
+      public HttpContext Context { get; }
+      public Hashtable Data { get; }
+    }
+
+    private const string c_dataDictionaryKey = "Rm.DD";
+
+    // Ignore the obsolete warnings for AsyncLocalStorageProvider
+#pragma warning disable 618
+    private readonly AsyncLocalStorageProvider _fallbackProvider = new();
+#pragma warning restore 618
+
+    private readonly ThreadLocal<WeakReference<WrappedHashtable?>> _dataTable = new(() => new WeakReference<WrappedHashtable?>(null));
```

Comment:
All the interactions with the cached value and the weak reference are getting difficult to follow. I think we should consider moving all of the logic into a struct which encapsulates the details of accessing the hashtable. Something like:

```c#
struct Wrapper
{
  private class WrappedHashtable { ... }

  private readonly ThreadLocal<WeakReference<WrappedHashtable?>> _dataTable;

  public bool TryGetHashtable(HttpContext context, out Hashtable result) { ... }

  public Hashtable GetOrCreateHashtable(HttpContext context) { ... }

  public Hashtable RemoveHashtable(HttpContext context) { ... }
}
```

This way all the caching logic is at one place and the actual safe context methods are streamlined. By providing the `HttpContext` as parameter we can make `WrappedHashtable` an implementation detail and simplify the outer code. Naming, maybe `CachedHashtableAccessor` or something?

What do you think?